baking/src/tezos_baking/tezos_setup_wizard.py

```diff
@@ -430,16 +432,131 @@ def hashes_comply(s1, s2):
                             and artifact["history_mode"] == "full"
                         )
                     )
-                    and hashes_comply(
-                        artifact["tezos_version"]["commit_info"]["commit_hash"],
-                        node_version_hash,
-                    ),
+                    and pred(artifact),
                     iter(snapshot_array),
                 ),
-                {"url": None, "block_hash": None},
+                None,
+            )
+
+        def get_artifact_version(artifact):
+            version = artifact["tezos_version"]["version"]
+            # there seem to be some inconsistency with that field in different providers
+            # so the only thing we check is if it's a string
+            additional_info = version["additional_info"]
+            return (
+                version["major"],
+                version["minor"],
+                None if type(additional_info) == str else additional_info["rc"],
+            )
+
+        def compose_pred(*preds):
+            return reduce(
+                lambda acc, x: lambda artifact: acc(artifact) and x(artifact), preds
+            )
+
+        def sum_pred(*preds):
+            return reduce(
+                lambda acc, x: lambda artifact: acc(artifact) or x(artifact), preds
+            )
+
+        node_version = get_node_version()
+        major_version, minor_version, rc_version = node_version
+
+        exact_version_pred = lambda artifact: node_version == get_artifact_version(
+            artifact
+        )
+
+        exact_major_version_pred = (
+            lambda artifact: major_version == get_artifact_version(artifact)[0]
+        )
+
+        exact_minor_version_pred = (
+            lambda artifact: minor_version == get_artifact_version(artifact)[1]
+        )
+
+        exact_rc_version_pred = (
+            lambda artifact: rc_version == get_artifact_version(artifact)[2]
+        )
+
+        less_minor_version_pred = (
+            lambda artifact: minor_version > get_artifact_version(artifact)[1]
+        )
```

Comment:
All these predicates are using (part of) the artifact's version, couldn't they take that as an argument instead of the artifact itself?

IOW can't `get_artifact_version` be called only once?

On a related note, I find this name quite confusing, `artifact_version` makes me think that this has the version of the artifact, but really this is the version of the node that was used to make the snapshot. I recommend renaming these.

Reply:
> All these predicates are using (part of) the artifact's version, couldn't they take that as an argument instead of the artifact itself?

They could, but actually then `get_artifact_version` would become unnecessary at all. That also means that the responsibility of that function will be delegated to the predicates itself.

Reply:
> I suppose you mean that we can call it inside find_snapshot and then call predicates with individual version parts

Not quite, I mean that we can call it inside `find_snapshot` and then call predicates with _all_ the version parts.

IOW I'd like to give in input to the predicates the entire output of  `get_artifact_version`

This should make it possible to also compose these predicates with one-another, no?

Reply:
Makes sense, this change is included in recent fixup